cluster-autoscaler/cloudprovider/azure/azure_cache.go

```diff
@@ -197,58 +207,87 @@ func (m *azureCache) regenerate() error {
 	return nil
 }
 
+// fetchAzureResources retrieves and updates the cached Azure resources.
+//
+// This function performs the following:
+// - Fetches and updates the list of Virtual Machine Scale Sets (VMSS) in the specified resource group.
+// - Fetches and updates the list of Virtual Machines (VMs) and identifies the node pools they belong to.
+// - Maintains a set of VMs pools and VMSS resources which helps the Cluster Autoscaler (CAS) operate on mixed node pools.
+//
+// Returns an error if any of the Azure API calls fail.
 func (m *azureCache) fetchAzureResources() error {
 	m.mutex.Lock()
 	defer m.mutex.Unlock()
 
-	switch m.vmType {
-	case vmTypeVMSS:
-		// List all VMSS in the RG.
-		vmssResult, err := m.fetchScaleSets()
-		if err == nil {
-			m.scaleSets = vmssResult
-		} else {
-			return err
-		}
-	case vmTypeStandard:
-		// List all VMs in the RG.
-		vmResult, err := m.fetchVirtualMachines()
-		if err == nil {
-			m.virtualMachines = vmResult
-		} else {
-			return err
-		}
+	// NOTE: this lists virtual machine scale sets, not virtual machine
+	// scale set instances
+	vmssResult, err := m.fetchScaleSets()
+	if err != nil {
+		return err
 	}
+	m.scaleSets = vmssResult
+	vmResult, vmsPoolSet, err := m.fetchVirtualMachines()
+	if err != nil {
+		return err
+	}
+	// we fetch both sets of resources since CAS may operate on mixed nodepools
+	m.virtualMachines = vmResult
+	m.vmsPoolSet = vmsPoolSet
```

Comment:
Should we do double (VM + VMSS fetches) all the time? 
But it is already merged it seems. Might need a bit of reevaluation if to be cherry-picked.

Reply:
This intersting part is already in, so safe